File: benchmark-java-sync/src/main/java/com/example/benchmarkjavasync/impl/ApplicationDao.java
package com.example.benchmarkjavasync.impl;

import com.example.benchmarkjavasync.core.ApplicationRepository;
import com.example.benchmarkjavasync.core.model.CreditInformation;
import com.example.benchmarkjavasync.core.model.Passport;
import com.example.benchmarkjavasync.core.model.User;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ApplicationDao implements ApplicationRepository<Optional<UUID>, Void> {

    private final JdbcTemplate template;

    @Override
    public Optional<UUID> saveUser(User user) {
        template.update("insert into a_user values (?, ?, ?)",
                user.id(),
                user.firstName(),
                user.lastName()
        );
        return Optional.of(user.id());
    }

    @Override
    public Void saveCreditInformation(CreditInformation creditInformation) {
        template.update("insert into credit_information values (?, ?, ?, ?)",
                creditInformation.id(),
                creditInformation.amount(),
                creditInformation.status(),
                creditInformation.userId()

        );
        return null;
    }

    @Override
    public Void savePassport(Passport passport) {
        template.update("insert into passport values (?, ?, ?, ?)",
                passport.id(),
                passport.serialNumber(),
                passport.number(),
                passport.userId()
        );
        return null;
    }
}

File: benchmark-kotlin-async/src/main/kotlin/com/example/benchmarkkotlinasync/impl/ApplicationDao.kt
package com.example.benchmarkkotlinasync.impl;

import com.example.benchmarkkotlinasync.core.ApplicationRepository
import com.example.benchmarkkotlinasync.core.model.CreditInformation
import com.example.benchmarkkotlinasync.core.model.Passport
import com.example.benchmarkkotlinasync.core.model.User
import org.springframework.r2dbc.core.DatabaseClient
import org.springframework.r2dbc.core.await
import org.springframework.stereotype.Service
import java.util.*

@Service
class ApplicationDao(private val client: DatabaseClient) : ApplicationRepository<UUID, Unit> {
    override suspend fun saveUser(user: User): UUID {
        client.sql("insert into a_user values (:id, :first_name, :last_name)")
            .bind("id", user.id)
            .bind("first_name", user.firstName)
            .bind("last_name", user.lastName)
            .await()
        return user.id
    }

    override suspend fun saveCreditInformation(creditInformation: CreditInformation) {
        client.sql("insert into credit_information values (:id, :amount, :status, :user_id)")
            .bind("id", creditInformation.id)
            .bind("amount", creditInformation.amount)
            .bind("status", creditInformation.status)
            .bind("user_id",creditInformation.userId)
            .await()
    }

    override suspend fun savePassport(passport: Passport) {
        client.sql("insert into passport values (:id, :serial_number, :number, :user_id)")
            .bind("id", passport.id)
            .bind("serial_number", passport.serialNumber)
            .bind("number", passport.number)
            .bind("user_id",passport.userId)
            .await()
    }
}
